daemon/&container/: enable `--security-opt writable-cgroups=true|false` as an option

vbatts opened the pull request:
Fixes #42040
Fixes #42275
Closes #42043

Rather than making cgroups read-write by default, instead have a flag for making it possible.

Since these security options are passed through the cli to daemon API, no changes are needed to docker-cli.

Since this is currently only a single toggle, I also considered making it a `bool` like `cgroups-rw=true`. I could go either way. It being a string makes the intuitive value kindof hidden to users even moreso than the args to --security-opt already are.

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

Summoned the strong desire to run systemd inside a container. Particularly a docker container, not podman.
Sought out what is blocking it.
Talked to fine humans like @tianon.

**- How I did it**

wrote the code.

**- How to verify it**

- build and run this branch of `dockerd`. Nothing special.
- run:
```shell
vbatts@jungle:~$ docker run -it --rm --security-opt cgroups=rw debian:testing findmnt | grep cgroup.*rw
| `-/sys/fs/cgroup   cgroup                          cgroup2 rw,nosuid,nodev,noe
```

Further, with a base container like:

```Dockerfile
FROM debian:testing
RUN apt update && \
    apt install -y --no-install-recommends systemd && \
    rm -rf /var/lib/apt/lists/*
ENTRYPOINT ["/usr/lib/systemd/systemd"]
```

![image](https://github.com/user-attachments/assets/23744369-77d5-42b6-ac49-faa26b08c55e)


**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section:
-->
```markdown changelog
 `POST /containers/create` now accepts a `writable-cgroups=true` option in `HostConfig.SecurityOpt`
  to mount the container's cgroups writable. This provides a more granular
  approach than `HostConfig.Privileged`.
```

**- A picture of a cute animal (not mandatory but encouraged)**

![image](https://github.com/user-attachments/assets/e2592645-6b3e-4939-8384-9d2727b47c69)

vbatts committed:
daemon/&container/: enable `--security-opt writable-cgroups=true` as an option

Fixes #42040
Closes #42043

Rather than making cgroups read-write by default, instead have a flag
for making it possible.

Since these security options are passed through the cli to daemon API,
no changes are needed to docker-cli.

Signed-off-by: Vincent Batts <vbatts@hashbangbash.com>

vbatts commented:
Tested the above with:
```shell
vbatts@jungle:~/src/moby$ docker version
Client:
 Version:           26.1.5+dfsg1
 API version:       1.45
 Go version:        go1.23.2
 Git commit:        a72d7cd
 Built:             Sun Oct 20 11:38:38 2024
 OS/Arch:           linux/amd64
 Context:           default

Server:
 Engine:
  Version:          library-import
  API version:      1.48 (minimum version 1.24)
  Go version:       go1.22.5
  Git commit:       library-import
  Built:            library-import
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v2.0.0
  GitCommit:        207ad711eabd375a01713109a8a197d197ff6542
 runc:
  Version:          1.2.0+dev
  GitCommit:        v1.2.0-44-gd9eecde9
 docker-init:
  Version:          0.19.0
  GitCommit:        
```

and 

```shell
vbatts@jungle:~$ docker version
Client:
 Version:           26.1.5+dfsg1
 API version:       1.45
 Go version:        go1.23.2
 Git commit:        a72d7cd
 Built:             Sun Oct 20 11:38:38 2024
 OS/Arch:           linux/amd64
 Context:           default

Server:
 Engine:
  Version:          library-import
  API version:      1.48 (minimum version 1.24)
  Go version:       go1.22.5
  Git commit:       library-import
  Built:            library-import
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.7.22~ds1
  GitCommit:        1.7.22~ds1-1+b1
 runc:
  Version:          1.1.12+ds1
  GitCommit:        1.1.12+ds1-5.1+b1
 docker-init:
  Version:          0.19.0
  GitCommit:        
```

thaJeztah commented on daemon/daemon_unix.go:
```diff
@@ -249,6 +249,11 @@ func parseSecurityOpt(securityOptions *container.SecurityOptions, config *contai
 				return fmt.Errorf("invalid --security-opt 2: %q", opt)
 			}
 			securityOptions.NoNewPrivileges = noNewPrivileges
+		case "cgroups":
```
As this introduces a new option accepted by the API, can you add a bullet to the API version-history for v1.48? https://github.com/moby/moby/blob/20062ef3580a15e53c48e8e0e5aa2c43eb4982f1/docs/api/version-history.md#v148-api-changes

Something like (better suggestions welcome!)

```markdown
* `POST /containers/create` now accepts a `cgroups=rw` option in `HostConfig.SecurityOpt`
  to mount the container's cgroups writable. This provides a more granular
  approach than `HostConfig.Privileged`.
```

We should also include these options in our swagger / OpenAPI, but looks like we currently don't, so that's something we can do in a follow-up (I can create a tracking ticket for that).

thaJeztah reviewed (commented):
Good seeing you, @vbatts!

OHMAN this one brought me down the rabit-hole a bit while looking at some parts of the code (bear with me!)

Some thoughts on the "API" impact on this, and some pretty random thoughts on a possible alternative implementation (but pretty much open to suggestions to all of this)

### Creating containers;

- We've started to be more strict on new features introduced in newer API version, to prevent things working on the _same_ API version, but older/newer version of the daemon
- Unfortunately, currently we don't have a good approach for these options (i.e., the code parsing these isn't aware of the API version); we could add something to the router and either explicitly _ignore_ or error if it's passed on older API versions, similar to https://github.com/moby/moby/blob/20062ef3580a15e53c48e8e0e5aa2c43eb4982f1/api/server/router/container/container_routes.go#L638-L644
- :point_up: we can probably do so in a follow up (not a hard blocker)

### Inspecting containers;

Running `docker inspect` uses `InspectResponse ` for the response; https://github.com/moby/moby/blob/20062ef3580a15e53c48e8e0e5aa2c43eb4982f1/api/types/container/container.go#L169

This PR currently doesn't include the new option (but will include the option in `HostConfig.SecurityOpt`); do we think it'd be good to also show that it's set as part of the `inspect` response?

### Alternative implementation 🎲 🎲 

This is a bit of a wild-card; while looking at this PR, I was considering; "don't we already have an option for this????" and... well, we do; https://github.com/moby/moby/blob/20062ef3580a15e53c48e8e0e5aa2c43eb4982f1/daemon/create_unix.go#L40-L43


But for some reason we don't use that options for either `/sys/fs/cgroup` (or `/sys`); for those we set the `ro` option as part of the mount itself https://github.com/moby/moby/blob/20062ef3580a15e53c48e8e0e5aa2c43eb4982f1/oci/defaults.go#L79-L90


Which got me thinking;

- Should we consider removing the `ro` option there,
- Instead add those to the [default set of `ReadOnlyPaths`](https://github.com/moby/moby/blob/20062ef3580a15e53c48e8e0e5aa2c43eb4982f1/oci/defaults.go#L118-L124)
- For `Privileged` we already clear `ReadOnlyPaths` (so the "special" handling of `cgroups` would no longer be needed)
- For this new option, we can remove `/sys/fs/cgroup` from the list of `ReadOnlyPaths`
- :point_up: in which case, I'm even wondering; do we need to store that `SecurityOpts` on disk for this? (not sure about that one!) 

Ironically, I also stumbled on an `WithWriteableCgroupfs` option provided by containerd that we currently don't use (that's more something we could consider for a follow-up though)
https://github.com/moby/moby/blob/20062ef3580a15e53c48e8e0e5aa2c43eb4982f1/vendor/github.com/containerd/containerd/oci/spec_opts.go#L1144-L1156

vbatts commented:
- re API gating, I can look into adding that check.
- re "inspect" output, like you said it **_is_** there in the `HostConfig`
[inspect.json](https://github.com/user-attachments/files/17649024/inspect.json)
- re alternatives:
  - There was palpable resistance to making cgroups rw by default in #42043. This would need further agreement and review of "how bad could it be??". 
  - as for making this more like a "default" behaviour, running systemd as pid 1 in the container is honestly what I _feel_ when I see the `--init` flag, despite _knowing_ that it's just for the subreaper pid cleanup nonsense. But this is not a hill I am caring to consider, much less die on.
  - regarding the other Privileged related ponderings, I have no idea. Seems like a different investigation. It would be not good to require users to use `--privileged` just to run systemd in the container though. 
- I'm flexible on the field name. 
- Will add the json tag stuffs
-

thaJeztah commented:
> There was palpable resistance to making cgroups rw by default

Yeah, the approach would still be to make it read-only by default, but instead of as part of the [default mounts that are defined](https://github.com/moby/moby/blob/20062ef3580a15e53c48e8e0e5aa2c43eb4982f1/oci/defaults.go#L60), the mount itself wouldn't have `ro`, but instead `/sys/fs/cgroup` (and `/sys`) would be added to the [`ReadOnlyPaths`](https://github.com/moby/moby/blob/20062ef3580a15e53c48e8e0e5aa2c43eb4982f1/oci/defaults.go#L118-L124) list.

I _THINK_ (but didn't try) that would have the same net-result, but using a different mechanism. But it was really thinking out loud here (also wondering _"why didn't we do that in the first place?"_).

My train of thought there was that both would end up in `HostConfig` on disk, but adding (or removing for the `rw` case) `/sys/fs/cgroup` to the `ReadonlyPaths` would mean that there's no schema changes needed in `HostConfig`.

thaJeztah commented:
> I _THINK_ (but didn't try) that would have the same net-result

Ignore me for now; I tried doing so, but looks like it doesn't work (but would have to dig "why")

thaJeztah commented:
> but that is not right as an older cli now fails to talk to this newer dockerd, even though it works fine without this check. I'm missing the goal here. I'm going to punt that gating.

Yes, that's the intent; a docker < v28 (which is the release this would go in) CLI only supports up until API version X, for example, a docker 27.3.1 supports API v1.47;

```
docker version
Client:
 Version:           26.1.4
 API version:       1.45
```

When connecting with the daemon, it negotiates the highest API version that's supported by both the client and the daemon (which can either be limited by the daemon (e.g. a v23.0 daemon would support anything up to API v1.42) or the client). If the client is older than the daemon, the API version used is downgraded, and only features available in that API version can be used.

For testing, you can override the API version, e.g. `DOCKER_API_VERSION=1.48 docker ........`, which makes the cli pretend it supports an API version it doesn't support.

vbatts commented:
still seems funny, but whatever.

![image](https://github.com/user-attachments/assets/c5c77d30-86da-4554-8679-896051952478)

AkihiroSuda commented:
`podman` and `nerdctl` have `--systemd=(true|false|always)` flag.

Maybe  `docker` should follow the same convention?

AkihiroSuda commented on daemon/oci_linux.go:
```diff
@@ -666,7 +666,7 @@ func withMounts(daemon *Daemon, daemonCfg *configStore, c *container.Container,
 		// TODO: until a kernel/mount solution exists for handling remount in a user namespace,
 		// we must clear the readonly flag for the cgroups mount (@mrunalp concurs)
-		if uidMap := daemon.idMapping.UIDMaps; uidMap != nil || c.HostConfig.Privileged {
+		if uidMap := daemon.idMapping.UIDMaps; uidMap != nil || c.HostConfig.Privileged || c.Cgroups == "rw" {
```
"rw" should be prohibited on cgroup v1 hosts, as the container can be escaped via a write access to the device controller

vbatts replied:
good catch. Will need to cover that situation.

tianon replied:
To put that another way, IMO the v1 concerns are more of a documentation problem than something we should block in the code.

AkihiroSuda replied:
> this has security implications on the current v2 implementation also, so why block it for v1?

Because the issue is more critical on v1.

tianon replied:
https://github.com/moby/moby/pull/42043#issuecomment-781990747

and sure, but `--privileged` does it, so this is more about having a way to get that bit of `--privileged` without the full hammer (https://github.com/moby/moby/issues/47663) than it is about babysitting a user who's explicitly asking for an insecure (or potentially insecure) configuration -- all the `--security-opt` settings are "user accepts the risk" type peeling back of the security boundary of the container, and it's already up to users to determine how that affects them, their use case, and their security model (and v1 is on the way out completely anyhow)

AkihiroSuda reviewed (dismissed):
"rw" should be prohibited on cgroup v1 hosts, as the container can be escaped via a write access to the device controller

vbatts commented:
> `podman` and `nerdctl` have `--systemd=(true|false|always)` flag.
> 
> Maybe `docker` should follow the same convention?

That would be amazing consistency to have. First step would be even the ability to run systemd though.

tianon commented:
I'm personally pretty strongly opposed to a `--systemd` flag, especially one that amounts to being just a helper for setting a couple of other flags that are otherwise unrelated to systemd except that it happens to be a bit of software that requires them (I believe one could accomplish that via the CDI integration and a CDI definition that specifies the required parameters in aggregate).

tianon commented:
I also don't understand the reason for blocking this for older clients -- it's not breaking functionality, and the default still matches the prior default, so it seems harmless for an older client to be able to specify it? :thinking:

cpuguy83 commented:
This looks like a good addition.
Only comment is to make the opt a boolean like `no-new-privileges`, e.g. `writeable-cgroups=true|false`

vbatts commented:
![image](https://github.com/user-attachments/assets/8b20eaad-60c7-42d6-81dd-6570ddb54d24)

vbatts commented:
#realtalk I think it is spelled "writable"  not "writeable"

vbatts commented:
![image](https://github.com/user-attachments/assets/784a3bdc-a384-4b7b-a349-658e7db0f3ee)

AkihiroSuda commented on container/container.go:
```diff
@@ -122,6 +122,7 @@ type SecurityOptions struct {
 	AppArmorProfile string
 	SeccompProfile  string
 	NoNewPrivileges bool
+	WritableCgroups bool `json:"writable-cgroups,omitempty"`
```
This json tag doesn't seem consistent with others

vbatts replied:
how come?

AkihiroSuda replied:
Other options in the struct do not have `json` tags

vbatts replied:
oh oh, you said json tag, and my brain read variable name.
yes. @thaJeztah said about having the json tag, but to smooth it out, would mean changes on lines that have nothing to do with this PR ... so I didn't touch them.

AkihiroSuda replied:
So, `json:"writable-cgroups,omitempty"` should be removed too?

vbatts replied:
I'm ok with that. Since I've heard both for and against, I would welcome a tie-breaker from other maintainers.

thaJeztah replied:
Sorry, I'm lagging behind on my notifications; yeah, so my comment was related to some situations whree we didn't define the casing to use for fields, and some `ID` field renamed (`Id -> ID`). Golang itself is (sigh) case-insensitive for JSON fields, so all seemed fine, but we broke some non-Go tools, so we now had to explicitly set `Id` after-the-fact to preserve the `Id` casing. So my thinking here was that it's better to be explicit instead of depending on Go's behavior. 

The `omitempty` was from a perspective that it's a new field, and adding the `omitempty` could possibly prevent container configs being updated on-disk during the "restore" cycle on daemon startup.

vbatts replied:
Should I add a commit that also tags the other fields as well?
(that feels like venturing into ensuring and fixing marshalling issues
that are unrelated to this PR...)

cpuguy83 commented on api/server/router/container/container_routes.go:
```diff
@@ -643,6 +643,20 @@ func (c *containerRouter) postContainersCreate(ctx context.Context, w http.Respo
 		}
 	}
 
+	if versions.LessThan(version, "1.48") {
```
Oh we also discussed not gating this on API.
@tianon @thaJeztah Did we come to a conclusion on this?

tianon replied:
I'm +1(00) on dropping the API version limitations (with thanks and apologies to vbatts for having implemented it successfully), but I think Seb still had some reservations unless I managed to be convincing enough. :bow: :heart:

vbatts commented:
I've stashed the API gating in a branch here: https://github.com/moby/moby/compare/master...vbatts:moby:cgroups-rw-api-gate
dropped that gating comit, and dropped the json tag so that the field is consistent.

PTAL

vbatts commented:
On 12/11/24 05:30 -0800, Sebastiaan van Stijn wrote:
>@thaJeztah commented on this pull request.

>Sorry, I'm lagging behind on my notifications; yeah, so my comment was related to some situations whree we didn't define the casing to use for fields, and some `ID` field renamed (`Id -> ID`). Golang itself is (sigh) case-insensitive for JSON fields, so all seemed fine, but we broke some non-Go tools, so we now had to explicitly set `Id` after-the-fact to preserve the `Id` casing. So my thinking here was that it's better to be explicit instead of depending on Go's behavior.
>
>The `omitempty` was from a perspective that it's a new field, and adding the `omitempty` could possibly prevent container configs being updated on-disk during the "restore" cycle on daemon startup.

Should I add a commit that also tags the other fields as well? 
(that feels like venturing into ensuring and fixing marshalling issues
that are unrelated to this PR...)

cpuguy83 reviewed (changes requested):
LGTM but needs an integration test in integration/

Thanks!

kolyshkin commented:
In podman, the relevant CLI option is `--security-opt=unmask=/sys/fs/cgroup`, described [here](https://docs.podman.io/en/v5.2.2/markdown/podman-run.1.html) (look for `unmask).

If at all possible, can the option being added be made compatible with podman?

tianon commented:
I'm not totally opposed, but it is definitely a weird way to say it -- `/sys/fs/cgroup` isn't currently "masked", it's just mounted read-only, so it's a really awkward UX for "unmask" to be the way to swap it to be writeable.

vbatts commented:
that seems like an open ended option to not mount over various paths. I get it that podman has that functionality before this PR, but it does seem a bit confusing behavior to expect end users to keep track of having to pass the path specifically. :-\

vbatts commented:
> On 13/11/24 07:18 -0800, Brian Goff wrote: @cpuguy83 requested changes on this pull request. LGTM but needs an integration test in integration/
> can you point to where similar `docker run --security-options ...` are being tested?

This is an honest request. 
I was looking at some of the tests in integration/ and not entirely grokking what I need to do here

cpuguy83 commented:
Sorry, this fell off my radar last week.
Here's an example of a security opt test: https://github.com/moby/moby/blob/54fff36d956a82fae2664b34038ae7cf778be04a/integration/container/run_linux_test.go#L358-L402

I think we can add the new test to the file.

vbatts commented:
Thanks for that. I tried a bit this weekend, and not sure if the
failures I'm seeing are due to the tests running inside docker or not. 
Haven't gotten to the bottom of it yet, but it was like the build of
dockerd didn't have my changeset at all.

cpuguy83 commented:
@vbatts If you want to push up your changes we can take a look.
Note that in order to run the test you need to execute `make test-integration`
You can set `TEST_FILTER=<go test filter>` to execute the individual test.

vbatts committed:
integration/container: test the writable-cgroups security option

Signed-off-by: Vincent Batts <vbatts@hashbangbash.com>

vbatts commented:
Thanks for that FILTER. I've pushed what I was working from.

cpuguy83 commented on integration/container/run_linux_test.go:
```diff
@@ -401,3 +401,52 @@ func TestSeccomp(t *testing.T) {
+	for _, tc := range testCases {
+		cID := container.Run(ctx, t, apiClient, tc.ops...)
+		res, err := container.Exec(ctx, apiClient, cID, []string{"/bin/true"})
+		assert.NilError(t, err)
```
This assert is failing for negative test cases.
Maybe only do this assert if `tc.expectedExitCode == 0`

vbatts replied:
ok. I went about it in a slightly deconstructed way.

kolyshkin commented:
> I'm not totally opposed, but it is definitely a weird way to say it -- `/sys/fs/cgroup` isn't currently "masked", it's just mounted read-only, so it's a really awkward UX for "unmask" to be the way to swap it to be writeable.

My guess it's a broad definition of "masking", which includes not only overmounts, but also read-only mounts.

Making the naming consistent (or at least adding a compatible alias) will improve interoperability.

vbatts commented:
@kolyshkin I get that interpretation of "masking" but is still confusing. Also, for a user to have to provide the explicit path is not great.

vbatts commented:
updated PTAL

cpuguy83 commented on integration/container/run_linux_test.go:
```diff
@@ -401,3 +401,48 @@ func TestSeccomp(t *testing.T) {
 	}
 }
+
+func TestCgroupRW(t *testing.T) {
```
For the cases that don't error out currently, can we validate that cgroups rw/ro respectively?

vbatts replied:
k, i've added that, but now am curious why the build of `dockerd` from this branch works for me on my machine with the command `docker run -it --rm --security-opt writable-cgroups=true debian mkdir /sys/fs/cgroup/foo`, but it fails in this test as permission denied read-only file system.

vbatts replied:
hrm. I'm puzzling on why this works on the host, but not in the tests...

cpuguy83 commented on integration/container/run_linux_test.go:
```diff
@@ -401,3 +401,62 @@ func TestSeccomp(t *testing.T) {
+	testCases := []testCase{
+		{
+			ops: nil,
+			// no err msg, so, success
```
Looks like this one needs `expectedExitCode: 1` since its using the default case which should be non-writable.

cpuguy83 commented on integration/container/run_linux_test.go:
```diff
@@ -401,3 +401,62 @@ func TestSeccomp(t *testing.T) {
+	ctx := setupTest(t)
+	apiClient := testEnv.APIClient()
+
+	type testCase struct {
```
Can you add a name to the test cases and run them as subtests?

tianon committed:
Update TestCgroupRW with subtests + nil case fix

Signed-off-by: Tianon Gravi <admwiggin@gmail.com>

tianon commented:
@vbatts I hope you don't mind; I took the liberty of pulling your change down locally, adding the subtests @cpuguy83 asked for, and then it was very obvious that the (nil) case he pointed out is why the test fails so I fixed that as well -- I'll look at applying the necessary rebase now

tianon commented:
Heh, tests fail on rootless, which I guess we maybe should've foreseen?

```
--- FAIL: TestCgroupRW (1.48s)
    --- FAIL: TestCgroupRW/nil (0.21s)
    --- PASS: TestCgroupRW/writable=true (0.20s)
    --- FAIL: TestCgroupRW/writable=false (0.19s)
    --- PASS: TestCgroupRW/writeable=true (0.01s)
    --- PASS: TestCgroupRW/writable=1 (0.20s)
    --- PASS: TestCgroupRW/writable=potato (0.01s)
...
=== Failed
=== FAIL: amd64.integration.container TestCgroupRW/nil (0.21s)
    run_linux_test.go:465: assertion failed: 1 (tc.expectedExitCode int) != 0 (res.ExitCode int)
    --- FAIL: TestCgroupRW/nil (0.21s)

=== FAIL: amd64.integration.container TestCgroupRW/writable=false (0.19s)
    run_linux_test.go:465: assertion failed: 1 (tc.expectedExitCode int) != 0 (res.ExitCode int)
    --- FAIL: TestCgroupRW/writable=false (0.19s)

=== FAIL: amd64.integration.container TestCgroupRW (1.48s)

DONE 352 tests, 28 skipped, 3 failures in 151.365s
```

tianon commented:
```go
		// TODO: until a kernel/mount solution exists for handling remount in a user namespace,
		// we must clear the readonly flag for the cgroups mount (@mrunalp concurs)
		if uidMap := daemon.idMapping.UIDMaps; uidMap != nil || c.HostConfig.Privileged || c.WritableCgroups {
```

So accordingly, should specifying this security-opt on rootless be an error for now (and an appropriate skip in the test)?  Or perhaps only trying to disable it explicitly should be an error on rootless?

tianon commented:
Oh, it fails on Ubuntu 20.04 generally, which is interesting:

```
--- FAIL: TestCgroupRW (1.40s)
    --- FAIL: TestCgroupRW/nil (0.18s)
    --- PASS: TestCgroupRW/writable=true (0.18s)
    --- FAIL: TestCgroupRW/writable=false (0.20s)
    --- PASS: TestCgroupRW/writeable=true (0.01s)
    --- PASS: TestCgroupRW/writable=1 (0.19s)
    --- PASS: TestCgroupRW/writable=potato (0.01s)
...
=== Failed
=== FAIL: amd64.integration.container TestCgroupRW/nil (0.18s)
    run_linux_test.go:465: assertion failed: 1 (tc.expectedExitCode int) != 0 (res.ExitCode int)
    --- FAIL: TestCgroupRW/nil (0.18s)

=== FAIL: amd64.integration.container TestCgroupRW/writable=false (0.20s)
    run_linux_test.go:465: assertion failed: 1 (tc.expectedExitCode int) != 0 (res.ExitCode int)
    --- FAIL: TestCgroupRW/writable=false (0.20s)

=== FAIL: amd64.integration.container TestCgroupRW (1.40s)

DONE 387 tests, 4 skipped, 3 failures in 165.689s
```

tianon commented:
> This looks like a good addition. Only comment is to make the opt a boolean like `no-new-privileges`, e.g. `writeable-cgroups=true|false`

While working on that, I tried to add support for `--security-opt writable-cgroups` (no `=true`, matching `no-new-privileges`), and ran into https://github.com/docker/cli/blob/f4a68da19595d64c50b0bbc2b1f15e645943ed82/cli/command/container/opts.go#L920-L926 and was sad (why do we validate in both daemon _and_ CLI??)

tianon commented:
Testing locally, I can reproduce the failures with user namespaces (`--userns-remap=default` on the daemon); everything is "permission denied" -- however, I can't reproduce with rootless (`mkdir /sys/fs/cgroup/foo` inside a privileged container on a rootless daemon works fine) :thinking:

tianon commented:
I'm going to swap the order of the test asserts so they test the output/stderr first to hopefully give us better clues than the exit code alone as to why this is failing.

tianon commented:
> I can't reproduce with rootless (`mkdir /sys/fs/cgroup/foo` inside a privileged container on a rootless daemon works fine) 🤔

I've got `WARNING: Running in rootless-mode without cgroups. Systemd is required to enable cgroups in rootless-mode.` in `docker info`, but the container appears to have cgroups and the mkdir works, so uh, I don't know where to go with that, but Ubuntu 20.04 I think might be cgroupsv1 (which the test as-written likely can't handle)?

tianon committed:
Adjust TestCgroupRW to test output before exit code

This should give us more useful test output on failures.

Signed-off-by: Tianon Gravi <admwiggin@gmail.com>

tianon commented:
Ok, that's better; rootless:

```
=== RUN   TestCgroupRW
=== RUN   TestCgroupRW/nil
=== RUN   TestCgroupRW/writable=true
    run_linux_test.go:469: assertion failed: 
        --- ←
        +++ →
        @@ -1,2 +1 @@
        -mkdir: can't create directory '/sys/fs/cgroup/foo': Permission denied
         
        
=== RUN   TestCgroupRW/writable=false
=== RUN   TestCgroupRW/writeable=true
=== RUN   TestCgroupRW/writable=1
    run_linux_test.go:469: assertion failed: 
        --- ←
        +++ →
        @@ -1,2 +1 @@
        -mkdir: can't create directory '/sys/fs/cgroup/foo': Permission denied
         
        
=== RUN   TestCgroupRW/writable=potato
--- FAIL: TestCgroupRW (1.21s)
    --- PASS: TestCgroupRW/nil (0.17s)
    --- FAIL: TestCgroupRW/writable=true (0.19s)
    --- PASS: TestCgroupRW/writable=false (0.17s)
    --- PASS: TestCgroupRW/writeable=true (0.00s)
    --- FAIL: TestCgroupRW/writable=1 (0.15s)
    --- PASS: TestCgroupRW/writable=potato (0.00s)
```

tianon commented:
Ubuntu 20.04:

```
=== RUN   TestCgroupRW
=== RUN   TestCgroupRW/nil
    run_linux_test.go:467: assertion failed: string "" does not contain "Read-only file system"
    run_linux_test.go:472: assertion failed: 1 (tc.expectedExitCode int) != 0 (res.ExitCode int)
=== RUN   TestCgroupRW/writable=true
=== RUN   TestCgroupRW/writable=false
    run_linux_test.go:467: assertion failed: string "" does not contain "Read-only file system"
    run_linux_test.go:472: assertion failed: 1 (tc.expectedExitCode int) != 0 (res.ExitCode int)
=== RUN   TestCgroupRW/writeable=true
=== RUN   TestCgroupRW/writable=1
=== RUN   TestCgroupRW/writable=potato
--- FAIL: TestCgroupRW (1.27s)
    --- FAIL: TestCgroupRW/nil (0.17s)
    --- PASS: TestCgroupRW/writable=true (0.19s)
    --- FAIL: TestCgroupRW/writable=false (0.17s)
    --- PASS: TestCgroupRW/writeable=true (0.00s)
    --- PASS: TestCgroupRW/writable=1 (0.16s)
    --- PASS: TestCgroupRW/writable=potato (0.00s)
```

I think this confirms my hunch, because on cgroupsv1 /sys/fs/cgroup would be a tmpfs so naturally writeable without error in all cases.

tianon committed:
Error on invalid requests for writable-cgroups

This makes `WritableCgroups` a pointer so we can error when it's specified in invalid configurations (both rootless and user namespaces).

Signed-off-by: Tianon Gravi <admwiggin@gmail.com>

tianon committed:
Add support for bare `writable-cgroups` security-opt

This requires changes in the CLI to support fully, but matches our other boolean option handling (`no-new-privileges`).

Signed-off-by: Tianon Gravi <admwiggin@gmail.com>

tianon committed:
Satisfy linter

    integration/container/run_linux_test.go:459:3: The copy of the 'for' variable "tc" can be deleted (Go 1.22+) (copyloopvar)

(thanks Go)

Signed-off-by: Tianon Gravi <admwiggin@gmail.com>

tianon committed:
Adjust test to support cgroupv1

This has to be careful to do something non-destructive, which this *should* be.

Signed-off-by: Tianon Gravi <admwiggin@gmail.com>

tianon commented:
It'll need a squash, but fingers crossed that finally does it. :see_no_evil:

tianon commented:
Ubuntu 20.04, grrrrrrrr

```
=== RUN   TestCgroupRW
=== RUN   TestCgroupRW/nil
    run_linux_test.go:483: assertion failed: 1 (tc.expectedExitCode int) != 0 (res.ExitCode int)
=== RUN   TestCgroupRW/writable
=== RUN   TestCgroupRW/writable=true
=== RUN   TestCgroupRW/writable=false
    run_linux_test.go:483: assertion failed: 1 (tc.expectedExitCode int) != 0 (res.ExitCode int)
=== RUN   TestCgroupRW/writeable=true
=== RUN   TestCgroupRW/writable=1
=== RUN   TestCgroupRW/writable=potato
--- FAIL: TestCgroupRW (1.75s)
    --- FAIL: TestCgroupRW/nil (0.18s)
    --- PASS: TestCgroupRW/writable (0.19s)
    --- PASS: TestCgroupRW/writable=true (0.20s)
    --- FAIL: TestCgroupRW/writable=false (0.18s)
    --- PASS: TestCgroupRW/writeable=true (0.01s)
    --- PASS: TestCgroupRW/writable=1 (0.18s)
    --- PASS: TestCgroupRW/writable=potato (0.01s)
```

hopefully it's something obvious (maybe something I'll see more clearly with fresh eyes tomorrow or someone else will before then)

https://github.com/moby/moby/blob/033f97519cded22411eb279a6866be12dd8c8745/contrib/check-config.sh#L153

Using this same test is probably what we need to do, and might even fix this, so I'm going to commit that now before I logoff in case it Does The Thing.

tianon commented:
Whoa, that really did it. :partying_face:

I guess we need to squash now and reset the CI? :see_no_evil:

tianon commented:
@vbatts I'd also love you to re-review what I've done to make sure I haven't totally ruined your intent :bow: :heart:

thaJeztah commented:
@AkihiroSuda PTAL as well (you have a "changes requested" pending)

vbatts commented:
On 16/01/25 17:19 -0800, Tianon Gravi wrote:
>Ubuntu 20.04, grrrrrrrr

yeah, so cgroups v1. Should we even support that or just stick to v2?

tianon commented:
https://github.com/moby/moby/pull/48828#discussion_r1833230967

(and it's working now, including the test)

thaJeztah commented:
Discussing in the maintainers call,  and while allowing this  for cgroups v1  might be less  secure; it's an opt-in feature, and cgroup  v1  should  be  going away at  some point, which means that less special  cases  makes that easier.

I think we  should bring this  in

thaJeztah reviewed (approved):
LGTM

AkihiroSuda commented:
LGTM

vbatts commented:
hah, I opened the tab this morning to review again and LGTM. That's the nudge that the cosmos needed. :-D

thaJeztah commented:
> hah, I opened the tab this morning to review again and LGTM. That's the nudge that the cosmos needed. :-D

It was good seeing you here!

thaJeztah commented:
Ha! Jinx https://github.com/docker/cli/pull/3936#issuecomment-1367407283 (so, yes, I :100: share that opinion)

>  (I'd love to get rid of some of that logic in the ParseSecurityOpts; I don't think it should all belong in that function, or even in the CLI code)

I think this was one of the cases where we either tried to repurpose/share code for "daemon" and "client", or (when the CLI was still in this repository), validation was added in the CLI, and (due to "integration-cli" tests highly depending on the CLI) nobody noticed validation was done in the CLI and missing in the API (but the code moved to the CLI when the repositories were split).

I think there's various other examples of this, and we need to work on those, but before ripping it ouf ot the CLI we need to double-check if corresponding validation exists on the daemon side (I think I spotted a few where that was not the case).

For validation in general, I once wrote some notes https://github.com/docker/cli/pull/1859 (highly WIP, but pushed it at the time to share with others that were working on an actual guide); 

> ## Validation
> 
> Docker uses a client/server architecture. As a consequence, the environment in
> which the Docker CLI runs may not match the environment of the _daemon_. For
> this reason, validation on the client-side should be limited, and deferred as
> much as possible to the daemon.
> 
> Offloading validation to the daemon prevents validation-rules from diverging
> between both, and prevents the similar code having to be maintained _twice_. Or
> (worse) having validation code in the client, but unhandled by the daemon.
> 
> In addition; do not expect what's valid to never change (what's invalid today,
> may be supported by the daemon, or the kernel, tomorrow).
> 
> As a rule of thumb: it's the CLI's responsibility to convert commands and arguments
> into API requests. If the CLI is able to handle a value, and turn it into an a
> valid API request (even if values in the request are invalid), that responsibility
> is met. If the request fails, it should handle the error, and (where  needed)
> present it to the user.

There's some nuances there, e.g.  if the client is aware of some options not  being supported on older APIs (and silently ignored), in which case we tend to add some guardrails.

vbatts commented:
> IMO a good follow up to make the experience even better would be to completely remove all of the cli-side validation in https://github.com/docker/cli/blob/f4a68da19595d64c50b0bbc2b1f15e645943ed82/cli/command/container/opts.go#L920-L926 (but I'm not a CLI maintainer; they may feel differently and might prefer to add this as yet another special case)

It still confuses me how these projects are separate ...
